test_rclcpp/test/test_executor.cpp

```diff
@@ -118,6 +121,84 @@ TEST(CLASSNAME(test_executor, RMW_IMPLEMENTATION), multiple_executors) {
   ASSERT_THROW(executor2.add_node(node1), std::runtime_error);
 }
 
+// Check that the executor is notified when a node adds a new timer, publisher, subscription,
+// service or client.
+TEST(CLASSNAME(test_executor, RMW_IMPLEMENTATION), notify) {
+  rclcpp::executors::SingleThreadedExecutor executor;
+  auto executor_spin_lambda = [&executor]() {
+      executor.spin();
+    };
+  auto node = rclcpp::Node::make_shared("test_executor_notify");
+  executor.add_node(node);
+  {
+    std::thread spin_thread(executor_spin_lambda);
+    bool timer_triggered = false;
+    auto timer = node->create_wall_timer(
+      1_ms,
+      [&executor, &timer_triggered](rclcpp::TimerBase & timer)
+    {
+      timer_triggered = true;
+      executor.cancel();
+      timer.cancel();
+    });
+
+    spin_thread.join();
```

Comment:
I'm not sure it's worth changing this test, but for the future you might consider using a future here rather than the `timer_triggered` boolean. Something like this:

```c++
std::thread spin_thread(executor_spin_lambda);
std::promise<void> timer_triggered_promise;
std::shared_future<void> timer_triggered_future(timer_triggered_promise.get_future());
auto timer = node->create_wall_timer(
  1_ms,
  [&executor, &timer_triggered_promise](rclcpp::TimerBase & timer)
{
  timer_triggered_promise.set_value();
  timer.cancel();
});

EXPECT_EQ(std::future_status::ready, timer_triggered_future.wait_for(100_ms));
executor.cancel();
spin_thread.join();
```

The advantage of this approach is that if the timer fails to trigger the test doesn't just hang, but instead it fails an expected check and the test can continue to try other things. Currently the only way this test fails is if the test timeout kills the test.